src/substrait/builders/plan.py

```diff
@@ -0,0 +1,294 @@
+        rel = stalg.Rel(
+            set=stalg.SetRel(
+                inputs=[plan.relations[-1].root.input for plan in bound_inputs], op=op
+            )
+        )
+
+        return stp.Plan(
+            relations=[
+                stp.PlanRel(
+                    root=stalg.RelRoot(input=rel, names=bound_inputs[0].relations[-1].root.names)
+                )
+            ],
+            **_merge_extensions(*bound_inputs),
+        )
+    
+    return resolve
+
+def fetch(plan: stp.Plan, 
+        offset: UnboundExtendedExpression, 
+        count: UnboundExtendedExpression) -> UnboundPlan:
+    def resolve(registry: ExtensionRegistry) -> stp.Plan:
+        bound_plan = plan if isinstance(plan, stp.Plan) else plan(registry)
+        ns = infer_plan_schema(bound_plan)
+        
+        bound_offset = offset(ns, registry)
+        bound_count = count(ns, registry)
+
+        rel = stalg.Rel(
+            fetch=stalg.FetchRel(
+                input=bound_plan.relations[-1].root.input,
+                offset_expr=bound_offset.referred_expr[0].expression,
+                count_expr=bound_count.referred_expr[0].expression
+            )
+        )
+
+        return stp.Plan(
+            relations=[
+                stp.PlanRel(
+                    root=stalg.RelRoot(input=rel, names=bound_plan.relations[-1].root.names)
+                )
+            ],
+            **_merge_extensions(bound_plan, bound_offset, bound_count),
+        )
+    
+    return resolve
+
+
+def join(
+    left: stp.Plan,
+    right: stp.Plan,
+    expression: UnboundExtendedExpression,
+    type: stalg.JoinRel.JoinType,
+) -> UnboundPlan:
+    def resolve(registry: ExtensionRegistry) -> stp.Plan:
+        bound_left = left if isinstance(left, stp.Plan) else left(registry)
+        bound_right = right if isinstance(right, stp.Plan) else right(registry)
+        left_ns = infer_plan_schema(bound_left)
+        right_ns = infer_plan_schema(bound_right)
+
+        ns = stt.NamedStruct(
+            struct=stt.Type.Struct(
+                types=list(left_ns.struct.types) + list(right_ns.struct.types),
+                nullability=stt.Type.Nullability.NULLABILITY_REQUIRED,
+            ),
+            names=list(left_ns.names) + list(right_ns.names),
+        )
+        bound_expression: stee.ExtendedExpression = expression(ns, registry)
+
+        rel = stalg.Rel(
+            join=stalg.JoinRel(
+                left=bound_left.relations[-1].root.input,
+                right=bound_right.relations[-1].root.input,
+                expression=bound_expression.referred_expr[0].expression,
+                type=type,
+            )
+        )
+
+        return stp.Plan(
+            relations=[stp.PlanRel(root=stalg.RelRoot(input=rel, names=ns.names))],
+            **_merge_extensions(bound_left, bound_right, bound_expression),
+        )
+    
+    return resolve
+
+def cross(
+    left: Union[stp.Plan, UnboundPlan],
```

Comment:
Why does this take either a Plan or an UnboundPlan where other relations don't?  Is it worth adding a type definition for this Union so that all uses are kept in sync?

Reply:
all of them should take in `Union[stp.Plan, UnboundPlan]`, I thought I made the changes everywhere, I'll fix it. type definition sounds like a good idea.